lib/THC/THCGeneral.c

```diff
@@ -42,21 +101,153 @@ void THCudaInit(THCState* state)
+      newStreams[dev][stream] = state->streamsPerDevice[dev][stream];
+    }
+
+    /* Allocate new streams */
+    for (int stream = state->numUserStreams + 1; stream <= numStreams; ++stream) {
+      newStreams[dev][stream] = NULL;
+      THCudaCheck(cudaStreamCreate(&newStreams[dev][stream]));
+    }
+  }
+
+  cudaStream_t** oldStreams = state->streamsPerDevice;
+  state->streamsPerDevice = newStreams;
+  state->numUserStreams = numStreams;
+
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    free(oldStreams[dev]);
+  }
+  free(oldStreams);
+
+  THCudaCheck(cudaSetDevice(prevDev));
+}
+
+int THCState_getNumStreams(THCState* state)
+{
+  return state->numUserStreams;
+}
+
+void THCState_resetStreams(THCState* state, int device)
+{
+  if (state->currentStream !=
+      state->streamsPerDevice[device][state->currentPerDeviceStream]) {
+    THError("Unexpected stream state");
+  }
+
+  /* Reallocate all streams for the current device; the 0 stream
+     doesn't need updating */
+  for (int dev = 0; dev < state->numDevices; ++dev) {
+    for (int stream = 1; stream <= state->numUserStreams; ++stream) {
+      THCudaCheck(cudaStreamCreate(&state->streamsPerDevice[dev][stream]));
+    }
+  }
+
+  state->currentStream =
+    state->streamsPerDevice[device][state->currentPerDeviceStream];
+}
+
+cudaStream_t THCState_getDeviceStream(THCState *state, int device, int stream)
+{
+  /* `device` is a CUDA index */
+  if (device >= state->numDevices || device < 0)
+  {
+    THError("%d is not a device", device + 1 /* back to Torch index */);
+  }
+
+  /* Stream 0 is the default stream, 1 ... `numUserStreams` are Torch streams */
+  if (stream > state->numUserStreams || stream < 0)
+  {
+    THError("%d is not a stream", stream);
+  }
+
+  return state->streamsPerDevice[device][stream];
+}
+
+cudaStream_t THCState_getCurrentStream(THCState *state)
+{
+  /* This is called at the point of kernel execution.
+     For some debugging code or improperly instrumented kernels,
+     `state` is null */
+  if (state) {
+    return state->currentStream;
+  } else {
+    /* assume default stream */
+    return NULL;
+  }
+}
+
+int THCState_getCurrentStreamIndex(THCState *state)
+{
+  return state->currentPerDeviceStream;
+}
+
+void THCState_setStream(THCState *state, int device, int stream)
+{
+  /* `device` is a CUDA index */
+  if (device >= state->numDevices || device < 0)
+  {
+    THError("%d is not a device", device + 1 /* back to Torch index */);
+  }
+
+  /* Stream 0 is the default stream, 1 ... `numUserStreams` are Torch streams */
+  if (stream > state->numUserStreams || stream < 0)
+  {
+    THError("%d is not a stream", stream);
+  }
+
+  state->currentStream = state->streamsPerDevice[device][stream];
+  state->currentPerDeviceStream = stream;
+  THCudaBlas_setStream(state, device, state->currentStream);
+}
+
+void THCState_setStreamForCurrentDevice(THCState *state, int stream)
+{
+  if (state->currentPerDeviceStream != stream)
```

Comment:
I'm wondering if we should set the stream regardless of whether it's already set. We do the same for the device. It's necessary when the same Lua state is operated on in different threads. CUDA state is typically per-thread, so if you set the device on thread 1 and then launch kernels from device 2, it will just use the default device. I'm guessing `cublasSetStream` works the same way?